src/roff/grog/grog.1.man
.TH GROG @MAN1EXT@ "@MDATE@" "Groff Version @VERSION@"
.SH NAME
grog \(em guess options for a following groff command
.\" The .SH was moved to this place in order to appease `apropos'.
.
.\" grog.man -> grog.1 - man page for grog (section 1)
.\" Source file position:  <groff_source_top>/src/roff/grog/grog.man
.\" Installed position:    <prefix>/share/man/man1/grog.1
.
.\" --------------------------------------------------------------------
.\" Legalese
.\" --------------------------------------------------------------------
.
.de co
Copyright \(co 1989-2015 Free Software Foundation, Inc.

Last update: 29 Jul 2015

This file is part of grog, which is part of groff, a free software
project.  You can redistribute it and/or modify it under the terms of
the GNU General Public License version 2 (GPL2) as published by the
Free Software Foundation.

groff is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.

The text for GPL2 is available in the internet at
.UR http://\:www.gnu.org/\:licenses/\:gpl-2.0.txt
GNU copyleft site
.UE .
..
.
.de au
Written by James Clark.

Maintained by
.MT wl@gnu.org
Werner Lemberg
.ME .

Rewritten and put under GPL by
.MT groff-bernd.warken-72@web.de
Bernd Warken
.ME .
..
.
.\" --------------------------------------------------------------------
.\" Characters
.\" --------------------------------------------------------------------
.
.\" Ellipsis ...
.ie t .ds EL \fS\N'188'\fP\"
.el .ds EL \&.\|.\|.\&\"
.\" called with \*(EL
.
.\" Bullet
.ie t .ds BU \(bu
.el .ds BU *
.\" used in `.IP \*(BU 2m' (former .Topic)
.
.
.\" --------------------------------------------------------------------
.SH SYNOPSIS
.\" --------------------------------------------------------------------
.
.SY grog
.OP \-C
.OP \-T device
.OP \-\-run
.OP \-\-warnings
.OP \-\-ligatures
.OP \& "\%groff\-option no arg"
.OP \& "\%groff\-option with arg \*(EL"
.OP \-\-
.OP \& "\%filespec \*(EL"
.YS
.
.BR "grog \-h " | " \-\-help"
.br
.BR "grog \-v " | " \-\-version"
.
.
.\" --------------------------------------------------------------------
.SH DESCRIPTION
.\" --------------------------------------------------------------------
.
.B grog
reads the input (file names or standard input) and guesses which of
the
.BR groff (@MAN1EXT@)
options are needed to perform the input with the
.B groff
program.
.
A suitable device is now always written as
.BI -T device
including the groff default as
.BR "-T ps" .
.
.
.P
The corresponding
.B groff
command is usually displayed in standard output.
.
With the option
.BR \-\-run ,
the generated line is output into standard error and the generated
.B groff
command is run on the
.IR "standard output" .
.
.BR groffer (@MAN1EXT@)
relies on a perfectly running
.BR groff (@MAN1EXT@).
.
.
.\" --------------------------------------------------------------------
.SH OPTIONS
.\" --------------------------------------------------------------------
.
The option
.B \-v
or
.B \-\-version
prints information on the version number.
.
Also
.B \-h
or
.B \-\-help
prints usage information.
.
Both of these options automatically end the
.B grog
program.
.
Other options are thenignored, and no
.B groff
command line is generated.
.
.
The following 3 options are the only
.B grog
options,
.
.TP
.B \-C
this option means enabling the
.I groff
compatibility mode, which is also transfered to the generated
.B groff
command line.
.
.TP
.B \-\-ligatures
this option forces to include the arguments
.B -P-y -PU
within the generated
.B groff
command line.
.
.TP
.B \-\-run
with this option, the command line is output at standard error and
then run on the computer.
.
.TP
.B \-\-warnings
with this option, some more warnings are output to standard error.
.
.
.P
All other specified short options (words starting with one minus
character
.BR \- )
are interpreted as
.B groff
options or option clusters with or without argument.
.
No space is allowed between options and their argument.
.
Except from the
.BI \-m arg
options, all options will be passed on, i.e.\& they are included
unchanged in the command for the output without effecting the work of
.BR grog .
.
.
.P
A
.I filespec
argument can either be the name of an existing file or a single minus
.B \-
to mean standard input.
.
If no
.I filespec
is specified standard input is read automatically.
.
.
.\" --------------------------------------------------------------------
.SH DETAILS
.\" --------------------------------------------------------------------
.
.B grog
reads all
.I filespec
parameters as a whole.
.
It tries to guess which of the following
.B groff
options are required for running the input under
.BR groff :
.BR \-e ,
.BR \-g ,
.BR \-G ,
.BR \-j ,
.BR \-J ,
.BR \-p ,
.BR \-R ,
.BR \-s ,
.BR \-t .
.BR \-man ,
.BR \-mdoc,
.BR \-mdoc-old,
.BR \-me ,
.BR \-mm ,
.BR \-mom ,
and
.BR \-ms .
.
.
.P
The guessed
.B groff
command including those options and the found
.I filespec
parameters is put on the standard output.
.
.
.P
It is possible to specify arbitrary
.B groff
options on the command line.
.
These are passed on the output without change, except for the
.BI \-m arg
options.
.
.
.P
The
.B groff
program has trouble when the wrong
.BI \-m arg
option or several of these options are specified.
.
In these cases,
.B grog
will print an error message and exit with an error code.
.
It is better to specify no
.BI \-m arg
option.
.
Because such an option is only accepted and passed when
.B grog
does not find any of these options or the same option is found.
.
.
.P
If several different
.BI \-m arg
options are found by
.B grog
an error message is produced and the program is terminated with an
error code.
.
But the output is written with the wrong options nevertheless.
.
.
.P
Remember that it is not necessary to determine a macro package.
.
A
.I roff
file can also be written in the
.I groff
language without any macro package.
.
.B grog
will produce an output without an
.BI \-m arg
option.
.
.
.P
As
.B groff
also works with pure text files without any
.I roff
requests,
.B grog
cannot be used to identify a file to be a
.I roff
file.
.
.
.P
The
.BR groffer  (@MAN1EXT@)
program heavily depends on a working
.BR grog .
.
.
.P
The
.B grog
source contains two files written in different programming languages:
.
.ft CB
grog.pl
.ft R
is the
.I Perl
version, while
.ft CB
grog.sh
.ft R
is a shell script using
.BR awk (1).
During the run of
.BR make (1),
it is determined whether the system contains a suitable version of
.BR perl (1).
If so,
.ft CB
grog.pl
.ft R
is transformed into
.BR grog ;
otherwise
.ft CB
grog.sh
.ft R
is used instead.
.
.
.\" --------------------------------------------------------------------
.SH EXAMPLES
.\" --------------------------------------------------------------------
.
.IP \*(BU 2m
Calling
.RS
.IP
.EX
grog meintro.me
.EE
.RE
.IP
results in
.RS
.EX
.IP
groff \-me meintro.me
.EE
.RE
.
.IP
So
.B grog
recognized that the file
.B meintro.me
is written with the
.B \-me
macro package.
.RE
.
.
.IP \*(BU 2m
On the other hand,
.RS
.IP
.EX
grog pic.ms
.EE
.RE
.
.IP
outputs
.
.RS
.IP
.EX
groff \-p \-t \-e \-ms pic.ms
.EE
.RE
.
.IP
Besides determining the macro package
.BR \-ms ,
.B grog
recognized that the file
.B pic.ms
additionally needs
.BR \-pte ,
the combination of
.B \-p
for
.IR pic ,
.B \-t
for
.IR tbl ,
and
.B \-e
for
.IR eqn .
.RE
.
.
.IP \*(BU 2m
If both of the former example files are combined by the command
.
.RS
.IP
.EX
grog meintro.me pic.ms
.EE
.RE
.
.IP
an error message is sent to standard error because
.B groff
cannot work with two different macro packages:
.
.RS
.IP
.ft CR
grog: error: there are several macro packages: \-me \-ms
.ft
.RE
.
.IP
Additionally the corresponding output with the wrong options is printed
to standard output:
.
.RS
.IP
.EX
groff \-pte \-me \-ms meintro.me pic.ms
.EE
.RE
.
.IP
But the program is terminated with an error code.
.
.
.IP \*(BU 2m
The call of
.
.RS
.IP
.EX
grog \-ksS \-Tdvi grnexmpl.g
.EE
.RE
.
.IP
contains several
.B groff
options that are just passed on the output without any interface to
.BR grog .
These are the option cluster
.B \-ksS
consisting of
.BR \-k ,
.BR \-s ,
and
.BR \-S ;
and the option
.B \-T
with argument
.BR dvi .
The output is
.
.RS
.IP
.EX
groff \-k \-s \-S \-Tdvi grnexmpl.g
.EE
.RE
.
.IP
so no additional option was added by
.BR grog .
As no option
.BI \-m arg
was found by
.B grog
this file does not use a macro package.
.
.
.\" --------------------------------------------------------------------
.SH "SEE ALSO"
.\" --------------------------------------------------------------------
.
.ad l
.
.BR groff (@MAN1EXT@),
.BR groffer (@MAN1EXT@)
.BR @g@troff (@MAN1EXT@),
.BR @g@tbl (@MAN1EXT@),
.BR @g@pic (@MAN1EXT@),
.BR @g@chem (@MAN1EXT@),
.BR @g@eqn (@MAN1EXT@),
.BR @g@refer (@MAN1EXT@),
.BR @g@grn (@MAN1EXT@),
.BR grap (1),
.BR @g@soelim (@MAN1EXT@)
.
.RS
Man-pages of section 1 can be viewed with either
.RS
\(Do
.BI man " name"
.RE
for text mode or
.RS
\(Do
.BI groffer " name"
.RE
for graphical mode (default is PDF mode).
.RE
.
.
.P
.BR groff_me (@MAN7EXT@),
.BR groff_ms (@MAN7EXT@),
.BR groff_mm (@MAN7EXT@),
.BR groff_mom (@MAN7EXT@),
.BR groff_man (@MAN7EXT@)
.
.RS
Man-pages of section 7 can be viewed with either with
.RS
\(Do
.BI "man 7 " name
.RE
for text mode or
.RS
\(Do
.BI "groffer 7 " name
.RE
for graphical mode (default is PDF mode).
.RE
.
.
.\" --------------------------------------------------------------------
.SH "COPYING"
.\" --------------------------------------------------------------------
.co
.\" --------------------------------------------------------------------
.SH "AUTHORS"
.\" --------------------------------------------------------------------
.au
.
.
.\" --------------------------------------------------------------------
.\" Emacs settings
.\" --------------------------------------------------------------------
.
.\" Local Variables:
.\" mode: nroff
.\" End:


ChangeLog
2015-08-27  Bernd Warken  <groff-bernd.warken-72@web.de>

	* tmac/pfdpic.tmac: Add file for macro PDFPIC.

	* tmac/tmac.am: Add new tmac file.

	* man/groff_tmac.5.man: Add definition for the new tmac.

2015-08-22  Bernd Warken  <groff-bernd.warken-72@web.de>

	* all man-pages *.man: Rename man-pages such that the man section
	number before the final `.man'.

	* all *.am: Include renaming.

	* Makefile.am: Reset `.man.n' by `.[157].man.n' with SUFFIXES
	`.[157].man .n'.

2015-08-20  Bernd Warken  <groff-bernd.warken-72@web.de>

	* Makefile.am: Fix definition of man[157]_MANS.

2015-08-13  Bernd Warken  <groff-bernd.warken-72@web.de>

	* all *.am: Correct Emacs mode `End:' instead of `End.'.

2015-08-09  werner LEMBERG  <wl@gnu.org>

	Revert previous commit, since it doesn't work correctly yet.

2015-08-07  Bernd Warken  <groff-bernd.warken-72@web.de>

	Add section name to all man-page names.

	* Makefile.am: In `install-prefix-man' and `uninstall-prefix-man'
	change `sed' commands.  Correct saving in man[157]_MANS.

	* all *.man in source tree: Rename `manpage.man' to
	`manpage.[157].man'.

	* all makefile.am in source tree: Add new man-page file name.

2015-08-05  Bernd Warken  <groff-bernd.warken-72@web.de>

	* Makefile.am, arch/mingw/mingw.am, arch/misc/misc.am,
	* contrib/eqn2graph/eqn2graph.am, contrib/grap2graph/grap2graph,
	* contrib/pic2graph/pic2graph.am, doc/doc.am,
	* font/devX100/devX100.am, font/devX100-12/devX100-12.am,
	* font/devX75/devX75.am, font/devX75-12/devX75-12.am,
	* font/devascii/devascii.am, font/devcp1047/devcp1047.am,
	* font/devdvi/devdvi.am, font/devhtml/devhtml.am,
	* font/devlatin1/devlatin1.am, font/devlbp/devlbp.am,
	* font/devlj4/devlj4.am, font/devpdf/devpdf.am,
	* font/devps/devps.am, font/devutf8/devutf8.am,
	* font/scripts/scripts.am, man/man.am,
	* src/devices/grodvi/grodvi.am, src/devices/grohtml/grohtml.am,
	* src/devices/grolbp/grolbp.am, src/devices/grolj4/grolj4.am,
	* src/devices/gropdf/gropdf.am, src/devices/grops/grops.am,
	* src/devices/grotty/grotty.am,
	* src/include/include.am, src/libs/libbib/libbib.am,
	* src/libs/libdriver/libdriver.am, src/libs/libgroff/libgroff.am,
	* src/libs/libxutil/libxutil.am, src/preproc/grn/grn.am,
	* src/preproc/html/html.am, src/preproc/pic/pic.am,
	* src/preproc/preconv/preconv.am, src/preproc/tbl/tbl.am
	* src/preproc/refer/refer.am, src/preproc/soelim/soelim.am,
	* src/roff/groff/groff.am, src/roff/grog/grog.am,
	* src/roff/nroff/nroff.am, src/roff/troff/troff.am,
	* src/utils/addftinfo/addftinfo.am,
	* src/utils/afmtodit/afmtodit.am, src/utils/hpftodit/hpftodit.am,
	* src/utils/indxbib/indxbib.am, src/utils/lkbib/lkbib.am,
	* src/utils/lookbib/lookbib.am, src/utils/pfbtops/pfbtops.am,
	* src/utils/tfmtodit/tfmtodit.am, src/utils/xtotroff/xtotroff.am,
	* tmac/tmac.am:
	Add `Last update'.  Setup Emacs mode.

2015-07-29  Bernd Warken  <groff-bernd.warken-72@web.de>

	* src/roff/grog/subs.pl: Add -T for all `grog' output, also using
	-Tps when no option -T is given in the arguments.

	* src/roff/grog/grog.man: Add information about -T output.

2015-07-23  Bernd Warken  <groff-bernd.warken-72@web.de>

	* man/groff.man: Add last update.  Add link on Wikipedia on groff.
	Add link of groff tutorial by Laha.

2015-07-19  Bernd Warken  <groff-bernd.warken-72@web.de>

	* src/roff/grog/*.pl: Add last update.

	* src/roff/grog/subs.pl: Remove call to Perl INC::System::Simple,
	use backticks instead.
	Define &err().
	Reorder some extensions.
	Distinguish between .TH at first command for man and later on for
	ms.

2015-07-01 Peter Bray <pdb_ml@yahoo.com.au>

	Fix Savannah bug #44784.

	Compatibility Mode rendering of groff manual pages

	* m4/groff.m4 (GROFF_WITH_COMPATIBILITY_WRAPPERS): New macro to
	implement 'configure' option
	'--with-compatibility-wrappers=<value>'.  With this option the
	user can control how groff compatibility wrappers for
	vendor-provided non-GNU macro sets are installed.

	* configure.ac: Use new macro GROFF_WITH_COMPATIBILITY_WRAPPERS.

	* Makefile.am: Define tmac_*_prefix if 'compatibility_wrappers' is
	set to 'yes'.

	* tmac/tmac.am: Use 'compatibility_wrappers' to control the
	installation:

	- If set to `no', the groff-provided macro set are installed in
	<prefix>/share/groff/<version>/tmac/ with the original macro set
	name.

	- If set to `yes', the groff-provided macro set are installed in
	<prefix>/share/groff/<version>/tmac/ with a prefix of
	$(tmac_prefix) (which is usually `g') and the generated
	compatibility wrappers are installed in
	<prefix>/lib/groff/site-tmac, with the original macro set name.

	- If set to `manual', the groff-provided macro set are installed
	in <prefix>/share/groff/<version>/tmac/ with the original macro
	set name and the generated compatibility wrappers are installed in
	<prefix>/lib/groff/site-tmac, with a name of
	<macro><wrapper_suffix> (which is usually `-os').

2015-06-17  Werner LEMBERG  <wl@gnu.org>

	* tmac/tty.tmac: Fix mapping of \[bu] character for -Tutf8.

	Problem reported by carsten.kunze@arcor.de.

2015-06-15  Deri James  <deri@chuzzlewit.myzen.co.uk>

	pdfmom should use $RT_SEP as path separator.

	* src/devices/gropdf/gropdf.am: Use @RT_SEP@ as path separator.

	* src/devices/gropdf/pdfmom.pl: As above but also fix typo in regex
	for grep.

2015-05-08  Eli Zaretskii  <eliz@gnu.org>

	[groff] Make relocation really work.

	See

	  https://lists.gnu.org/archive/html/groff/2015-05/msg00000.html

	* src/roff/groff/groff.cpp: Include `relocate.h'.
	(main): Use `relocatep'.

2015-05-03  Bertrand Garrigues  <bertrand.garrigues@laposte.net>

	Fix Savannah bug #44768 (Solaris 10 compatibility issues).

	* m4/groff.m4 (GROFF_DIFF_D): new macro to check if `diff' has
	option -D, and search for `gdiff' if not (new variable DIFF_PROG
	will contain the `diff' program to be used).

	* m4/groff.m4 (GROFF_HAVE_TEST_EF_OPTION): new macro to check if
	`test' supports the option -ef (created new variable
	HAVE_TEST_EF_OPTION).


src/roff/grog/grog.pl
#! /usr/bin/env perl
# grog - guess options for groff command
# Inspired by doctype script in Kernighan & Pike, Unix Programming
# Environment, pp 306-8.

# Source file position: <groff-source>/src/roff/grog/grog.pl
# Installed position: <prefix>/bin/grog

# Copyright (C) 1993-2015  Free Software Foundation, Inc.
# Written by James Clark, maintained by Werner Lemberg.
# Rewritten with Perl by Bernd Warken <groff-bernd.warken-72@web.de>.
# The macros for identifying the devices were taken from Ralph
# Corderoy's `grog.sh' of 2006.

# Last update: 19 Jul 2015

# This file is part of `grog', which is part of `groff'.

# `groff' is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.

# `groff' is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see
# <http://www.gnu.org/licenses/gpl-2.0.html>.

########################################################################

require v5.6;

use warnings;
use strict;

# $Bin is the directory where this script is located
use FindBin;

my $before_make;	# script before run of `make'
{
  my $at = '@';
  $before_make = 1 if '@VERSION@' eq "${at}VERSION${at}";
}


our %at_at;
my $grog_dir;

if ($before_make) { # before installation
  my $grog_source_dir = $FindBin::Bin;
  $at_at{'BINDIR'} = $grog_source_dir;
# $grog_dir = $grog_source_dir;
  $grog_dir = '.';
  my $top = $grog_source_dir . '/../../../';
  open FILE, '<', $top . 'VERSION' ||
    die 'Could not open top file VERSION.';
  my $version = <FILE>;
  chomp $version;
  close FILE;
  open FILE, '<', $top . 'REVISION' ||
    die 'Could not open top file REVISION.';
  my $revision = <FILE>;
  chomp $revision;
  $at_at{'GROFF_VERSION'} = $version . '.' . $revision;
} else { # after installation}
  $at_at{'GROFF_VERSION'} = '@VERSION@';
  $at_at{'BINDIR'} = '@BINDIR@';
  $grog_dir = '@grog_dir@';
} # before make

die "$grog_dir is not an existing directory;" unless -d $grog_dir;


#############
# import subs

unshift(@INC, $grog_dir);
require 'subs.pl';


##########
# run subs

&handle_args();
&handle_file_ext(); # see $tmac_ext for gotten value
&handle_whole_files();
&make_groff_device();
&make_groff_preproc();
&make_groff_tmac_man_ms() || &make_groff_tmac_others();
&make_groff_line_rest();


1;
########################################################################
### Emacs settings
# Local Variables:
# mode: CPerl
# End:
